### ISSUE TITLE
Producer may create batches larger than broker will accept

### ISSUE THREAD

**Issue by yungchin** [Parsely/pykafka#424]
Just encountered this twice during benchmarking. I'm not sure what's particular about the situation yet, will have to dig for more details. I'm not completely clear yet that allowing this error to bubble up is the wrong behaviour: I suppose it depends on why the reset happens. Will try to find that out.

```
/srv/pykafka/benchmark/__init__.py in mk_topic(kafka_conn, num_partitions, replication_factor, msg_size_bytes, msg_sum_bytes, max_num_msgs)
    237         with topic.get_producer() as prod:
    238             for _ in xrange(n_msgs):
--> 239                 prod.produce(msg_size_bytes * b" ")
    240     return topic_name
    241 
/srv/pykafka/pykafka/producer.pyc in __exit__(self, exc_type, exc_value, traceback)
    182     def __exit__(self, exc_type, exc_value, traceback):
    183         """Context manager exit point - stop the producer"""
--> 184         self.stop()
    185 
    186     def start(self):
/srv/pykafka/pykafka/producer.pyc in stop(self)
    232         """Mark the producer as stopped"""
    233         self._running = False
--> 234         self._wait_all()
    235         if self._owned_brokers is not None:
    236             for owned_broker in self._owned_brokers.values():
/srv/pykafka/pykafka/producer.pyc in _wait_all(self)
    396         while any(q.message_is_pending() for q in itervalues(self._owned_brokers)):
    397             time.sleep(.3)
--> 398             self._raise_worker_exceptions()
    399 
    400 
/srv/pykafka/pykafka/producer.pyc in _raise_worker_exceptions(self)
    167                 log.error("Exception encountered in worker thread:\n%s",
    168                           "".join(traceback.format_tb(tb)))
--> 169             raise ex
    170 
    171     def __repr__(self):
error: [Errno 104] Connection reset by peer
```

**Comment by yungchin** [comment]
Ok, I think I figured out what the problem is here: the resets only occur if I'm hammering the topic with messages above a certain size each (it's ok for messages of 1kB, but will always crash if they're 10kB). At larger sizes, the producer worker will eventually build a `MessageSet` that the broker deems too large.

It's not possible to fix this by dialling down the `min_queued_messages` parameter (I went all the way down to `10`). The problem seems to me that the worker thread will always grab everything that's in the queue, so even if it triggers at 10 messages, when the benchmark thread is hammering the queue at the same time, the worker will eventually end up grabbing a batch of a thousand.

**Comment by emmett9001** [comment]
Does the `message.max.bytes` broker config option correlate to the behavior you're seeing? How about `socket.receive.buffer.bytes`, `socket.request.max.bytes`, or `socket.send.buffer.bytes`? Presumably we can have pykafka's default maximum batch size correspond to the appropriate one of these settings.

**Comment by emmett9001** [comment]
Possibly related to #438

**Comment by yungchin** [comment]
This should be fixed by #480.

**Comment by emmett9001** [comment]
#480 has been merged and put into a release, so this is fixed.
